### ISSUE TITLE
mapped memory crash when capturing in zink

### ISSUE THREAD

**Issue by zmike** [LunarG/gfxreconstruct#629]
I'm not entirely sure what's going on here, but when attempting to capture certain apps/tests, I'll randomly get crashes when attempting to write to a mapped buffer region. The apps/tests work fine without gfxr, so I think this is somehow a bug in the layer.

My latest example is trying to run GL CTS on RADV with:
`VK_INSTANCE_LAYERS=VK_LAYER_LUNARG_gfxreconstruct MESA_LOADER_DRIVER_OVERRIDE=zink ./glcts -n dEQP-GLES31.functional.shaders.opaque_type_indexing.sampler.dynamically_uniform.geometry.samplercubearray`

I map a memory region of size 1048576, then try to memcpy 1 byte into it, and it crashes.

Let me know what other info I can provide.

**Comment by zmike** [comment]
What's especially odd here is that I can run other cases (e.g., `dEQP-GLES31.functional.shaders.arrays_of_arrays.es31.constructor.implicit.mat3_3x3x3_fragment`) and they don't crash.

**Comment by bradgrantham-lunarg** [comment]
@zmike Do you get this same crash without #624?  (On systems without glibc 2.34, I mean).

**Comment by zmike** [comment]
I have no systems with a different glibc version, but I've been getting this for a while and only just got annoyed enough to report it :)

**Comment by bradgrantham-lunarg** [comment]
I want to differentiate between with PR #624 and without it, just in case that PR broke capture.  :)  So you're saying that you had this problem before my PR, correct?

**Comment by zmike** [comment]
That's correct.

**Comment by zmike** [comment]
Still broken as of today.

**Comment by bradgrantham-lunarg** [comment]
We're going to try to dive into this shortly.  We've needed someone with some facility building GL CTS and they're coming available.

Have you tried any of the other `GFXRECON_MEMORY_TRACKING_MODE`s or tried any other `PAGE_GUARD` settings?

**Comment by zmike** [comment]
I haven't. Given my lack of luck running gfxrecon to date, I'm afraid I don't know much about using it.

**Comment by zmike** [comment]
The crash doesn't actually require GLCTS though. It crashes when running pretty much anything under zink. I just crashed it the same way in supertuxkart.

**Comment by bradgrantham-lunarg** [comment]
That's very helpful, thanks.  We can try supertuxkart.  Any chance you've seen it crash in `gears`?

**Comment by zmike** [comment]
Yup, crashes in glxgears too.

**Comment by panos-lunarg** [comment]
I couldn't reproduce the issue with glxgears.
I'm not familiar with the zink internals so I'm not sure if it makes any difference switching the vulkan drivers between the open source and the proprietary one (I assume it does make a difference as I see the driver gets reported differently).

With both drivers glxgears gets captured, or at least a `gfxrecon_capture_` file gets generated, but fails to reproduce anything meaningful as `gfxrecon-play` tools exits with the message `File did not contain any frames`.
Converting the trace to ascii with the `gfxrecon-toascii` tool this makes sense as I couldn't find any calls to `vkQueuePresent()`

I managed to reproduce the segfault with the deqp test. Need to investigate further on that

**Comment by bradgrantham-lunarg** [comment]
Just out of curiosity, how does Zink present?

**Comment by zmike** [comment]
As of Mesa 22.1 it uses Vulkan WSI.

**Comment by panos-lunarg** [comment]
Apologies for the delay on this one. I think I have figured out what the problem is in this case.
Mesa apparently is creating worker threads where it delegates work like writing to BOs, or in our case, device memory mapped to user space. Running the deqp tests with `strace` and grepping for `sigprocmask` I got the following result:
```
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[SYS RTMIN RT_1], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[SYS RTMIN RT_1], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[SYS RTMIN RT_1], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[SYS RTMIN RT_1], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[SYS RTMIN RT_1], [], 8) = 0
...
Repeats several times
```
So it seams that apart from a couple of signals, all other signals are being blocked.
Grepping in the mesa source for `pthread_sigmask` yielded the function responsible for creating the mesa threads and for the observed behavior:
```
sigfillset(&new_set);
sigdelset(&new_set, SIGSYS);
pthread_sigmask(SIG_BLOCK, &new_set, &saved_set);
ret = thrd_create( &thread, routine, param );
pthread_sigmask(SIG_SETMASK, &saved_set, NULL);
```

Created a tentative PR #728 which essentially unblocks `SIGSEGV` each time GFXR applies some memory protection on a shadow memory. Perhaps this is can be too intrusive so it is open for discussion

**Comment by bradgrantham-lunarg** [comment]
@zmike is there a chance you can try Panos' branch in #728 with new option `GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV=true" and let us know if that solves your crash?

**Comment by zmike** [comment]
Sure thing, I'll test it out in the next day or two.

**Comment by zmike** [comment]
Sorry for the delay, the original mail reply must have gotten lost in my inbox. Using this mode, I now get this backtrace:

```
0x00007fffed5b776d in gfxrecon::encode::EncodePNextStruct(gfxrecon::encode::ParameterEncoder*, void const*) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
Missing separate debuginfos, use: dnf debuginfo-install expat-2.4.1-2.fc35.x86_64 glibc-2.34-11.fc35.x86_64 libICE-1.0.10-7.fc35.x86_64 libSM-1.2.3-9.fc35.x86_64 libXau-1.0.9-7.fc35.x86_64 libXdmcp-1.1.3-7.fc35.x86_64 libXext-1.3.4-7.fc35.x86_64 libXfixes-6.0.0-2.fc35.x86_64 libXxf86vm-1.1.4-17.fc35.x86_64 libedit-3.1-40.20210910cvs.fc35.x86_64 libffi-3.1-29.fc35.x86_64 libgcc-11.2.1-7.fc35.x86_64 libpng-1.6.37-11.fc35.x86_64 libstdc++-11.2.1-7.fc35.x86_64 libunwind-1.5.0-1.fc35.x86_64 libuuid-2.37.2-1.fc35.x86_64 libwayland-client-1.19.0-2.fc35.x86_64 libxshmfence-1.3-9.fc35.x86_64 libzstd-1.5.1-3.fc35.x86_64 llvm-libs-13.0.0-4.fc35.x86_64 ncurses-libs-6.2-8.20210508.fc35.x86_64 systemd-libs-249.7-2.fc35.x86_64 vulkan-loader-1.3.204.0-1.fc36.x86_64 zlib-1.2.11-30.fc35.x86_64
(gdb) bt
#0  0x00007fffed5b776d in gfxrecon::encode::EncodePNextStruct(gfxrecon::encode::ParameterEncoder*, void const*) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
#1  0x00007fffed5b8169 in gfxrecon::encode::EncodePNextStruct(gfxrecon::encode::ParameterEncoder*, void const*) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
#2  0x00007fffed5b8169 in gfxrecon::encode::EncodePNextStruct(gfxrecon::encode::ParameterEncoder*, void const*) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
#3  0x00007fffed6181c1 in gfxrecon::encode::EncodeStruct(gfxrecon::encode::ParameterEncoder*, VkFormatProperties2 const&) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
#4  0x00007fffed5c6983 in gfxrecon::encode::GetPhysicalDeviceFormatProperties2KHR(VkPhysicalDevice_T*, VkFormat, VkFormatProperties2*) () from /usr/local/lib64/libVkLayer_gfxreconstruct.so
#5  0x00007ffff6f99b52 in populate_format_props (screen=0x3917d80) at ../src/gallium/drivers/zink/zink_screen.c:1723
#6  0x00007ffff6f9b7bd in zink_internal_create_screen (config=0x7fffffffbab0) at ../src/gallium/drivers/zink/zink_screen.c:2286
#7  0x00007ffff6f9bcf5 in zink_create_screen (winsys=0x39117d0, config=0x7fffffffbab0) at ../src/gallium/drivers/zink/zink_screen.c:2368
#8  0x00007ffff59e6957 in sw_screen_create_zink (winsys=0x39117d0, config=0x7fffffffbab0, whatever=false) at ../src/gallium/auxiliary/target-helpers/sw_helper.h:117
#9  0x00007ffff655c723 in pipe_loader_sw_create_screen (dev=0x3911760, config=0x7fffffffbab0, sw_vk=false) at ../src/gallium/auxiliary/pipe-loader/pipe_loader_sw.c:425
#10 0x00007ffff655c14c in pipe_loader_create_screen_vk (dev=0x3911760, sw_vk=false) at ../src/gallium/auxiliary/pipe-loader/pipe_loader.c:171
#11 0x00007ffff655c16b in pipe_loader_create_screen (dev=0x3911760) at ../src/gallium/auxiliary/pipe-loader/pipe_loader.c:177
#12 0x00007ffff59f500d in kopper_init_screen (sPriv=0x390e2b0) at ../src/gallium/frontends/dri/kopper.c:159
#13 0x00007ffff59ec1bf in driCreateNewScreen2 (scrn=0, fd=-1, extensions=0x7ffff7fc0840 <kopper_extensions_noshm>, driver_extensions=0x7ffff767a680 <galliumvk_driver_extensions>, driver_configs=0x7fffffffc5a8, data=0x3837090) at ../src/gallium/frontends/dri/dri_util.c:142
#14 0x00007ffff59ec466 in driSWRastCreateNewScreen2 (scrn=0, extensions=0x7ffff7fc0840 <kopper_extensions_noshm>, driver_extensions=0x7ffff767a680 <galliumvk_driver_extensions>, driver_configs=0x7fffffffc5a8, data=0x3837090) at ../src/gallium/frontends/dri/dri_util.c:213
#15 0x00007ffff7f76665 in driswCreateScreenDriver (screen=0, priv=0x3815f80, driver=0x7ffff7fb7ba0 "zink") at ../src/glx/drisw_glx.c:963
#16 0x00007ffff7f769ce in driswCreateScreen (screen=0, priv=0x3815f80) at ../src/glx/drisw_glx.c:1040
#17 0x00007ffff7f7def2 in AllocAndFetchScreenConfigs (dpy=0x3818930, priv=0x3815f80) at ../src/glx/glxext.c:851
#18 0x00007ffff7f7e33f in __glXInitialize (dpy=0x3818930) at ../src/glx/glxext.c:976
#19 0x00007ffff7f77a7e in glXQueryVersion (dpy=0x3818930, major=0x381fff0, minor=0x381fff4) at ../src/glx/glxcmds.c:537
#20 0x000000000020d893 in tcu::lnx::x11::glx::GlxDisplay::GlxDisplay(tcu::lnx::EventState&, char const*) ()
#21 0x000000000020f3e6 in tcu::lnx::x11::glx::GlxRenderContext::GlxRenderContext(tcu::lnx::x11::glx::GlxContextFactory const&, glu::RenderConfig const&, glu::RenderContext const*) ()
#22 0x000000000020ff32 in tcu::lnx::x11::glx::GlxContextFactory::createContext(glu::RenderConfig const&, tcu::CommandLine const&, glu::RenderContext const*) const ()
#23 0x0000000000e49dc3 in deqp::Context::Context(tcu::TestContext&, glu::ContextType) ()
#24 0x0000000000f1f727 in deqp::TestPackage::init() ()
#25 0x000000000021ad0d in glcts::SingleConfigGL43TestPackage::init() ()
#26 0x000000000021ad5d in glcts::SingleConfigGL44TestPackage::init() ()
#27 0x000000000021adad in glcts::SingleConfigGL45TestPackage::init() ()
#28 0x00000000025019b5 in tcu::TestHierarchyIterator::next() ()
#29 0x00000000024b5780 in tcu::TestSessionExecutor::iterate() ()
#30 0x000000000247ed99 in tcu::App::iterate() ()
#31 0x000000000020aa70 in main ()
```

from `GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV=true VK_INSTANCE_LAYERS=VK_LAYER_LUNARG_gfxreconstruct LIBGL_ALWAYS_SOFTWARE=1 MESA_LOADER_DRIVER_OVERRIDE=zink ./glcts -n 'KHR-Single-GL46.enhanced_layouts.glsl_constant_integral_expression'`

If I've missed something about how this is intended to be used let me know, but I get this (new) crash from every app I attempt to run.

**Comment by panos-lunarg** [comment]
Hi Mike.

With a quick try I made I couldn't reproduce the crash. The `LIBGL_ALWAYS_SOFTWARE=1` variable got me a bit confused as when I set it to 1, gfxrecon does not get involved at all. Setting it to 0 or not setting it at all allows gfxrecon to get involved.

I think if you can share exact revisions would be helpful. Gfxrecon revision (parent on which you applied #728), mesa revision and deqp revision.
From the backtrace it looks like you have compiled deqp with `DEQP_TARGET=x11_glx`, is that correct?

I do get a few relevant looking warnings:
```
$ GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV=true VK_INSTANCE_LAYERS=VK_LAYER_LUNARG_gfxreconstruct MESA_LOADER_DRIVER_OVERRIDE=zink ./glcts -n 'KHR-Single-GL46.enhanced_layouts.glsl_constant_integral_expression'
Writing test log into TestResults.qpa
dEQP Core git-e22ec5fc4cbf4cbba97601433f1c93fb893a200d (0xe22ec5fc) starting..
  target implementation = 'X11 GLX'
[gfxrecon] INFO - Settings Loader: Found option "GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV" with value "true"
[gfxrecon] INFO - Initializing GFXReconstruct capture layer
[gfxrecon] INFO -   GFXReconstruct Version 0.9.14-dev (panos_629:744cb02*)
[gfxrecon] INFO - Recording graphics API capture to gfxrecon_capture_20220629T153656.gfxr
[gfxrecon] INFO -   Vulkan Header Version 1.3.216
[gfxrecon] WARNING - The application has specified that it uses Vulkan API version 1.3.218, which is newer than the version supported by GFXReconstruct.  Use of unsupported Vulkan features may cause capture or replay to fail.
WARNING: radv is not a conformant Vulkan implementation, testing use only.
WARNING: Some incorrect rendering might occur because the selected Vulkan device (AMD RADV SIENNA_CICHLID) doesn't support base Zink requirements: feats.features.alphaToOne line_rast_feats.rectangularLines line_rast_feats.smoothLines line_rast_feats.stippledRectangularLines line_rast_feats.stippledSmoothLines 
[gfxrecon] WARNING - Failed to copy entire pNext chain when unwrapping handles due to unrecognized sType 1000001002
[gfxrecon] WARNING - A pNext value with unrecognized VkStructureType = 1000001002 was omitted from the capture file, which may cause replay to fail.
[gfxrecon] WARNING - Failed to copy entire pNext chain when unwrapping handles due to unrecognized sType 1000001003
[gfxrecon] WARNING - A pNext value with unrecognized VkStructureType = 1000001003 was omitted from the capture file, which may cause replay to fail.

Test case 'KHR-Single-GL46.enhanced_layouts.glsl_constant_integral_expression'..
  Pass (Pass)

DONE!

Test run totals:
  Passed:        1/1 (100.0%)
  Failed:        0/1 (0.0%)
  Not supported: 0/1 (0.0%)
  Warnings:      0/1 (0.0%)
  Waived:        0/1 (0.0%)
```

In any case this looks like something different than the initial problem, so perhaps it would be preferable to file a new issue.

**Comment by andrew-lunarg** [comment]
FYI, alternative incantation to get software GL selected on platforms that have the `mesa_device_select` layer on implicitly:

`
    MESA_VK_DEVICE_SELECT="10005:0"
    MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1
```

`"10005:0"` Does the job for me but to get what the software Vk is called locally, run some Vulkan app once with this in the environment:

```
    MESA_VK_DEVICE_SELECT="list"
```
which shows list of available GPUs.

**Comment by panos-lunarg** [comment]
Hi @zmike 
Can you please verify that the crash related to llvmpipe is still reproducible? For me it seems that it is working:
```
GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV=true VK_INSTANCE_LAYERS=VK_LAYER_LUNARG_gfxreconstruct LIBGL_ALWAYS_SOFTWARE=1 MESA_LOADER_DRIVER_OVERRIDE=zink ./glcts -n 'KHR-Single-GL46.enhanced_layouts.glsl_constant_integral_expression'
Writing test log into TestResults.qpa
dEQP Core git-ef9a6635208b29dee3d79b2d8cf65294b48c6297 (0xef9a6635) starting..
  target implementation = 'X11 GLX'
[gfxrecon] INFO - Settings Loader: Found option "GFXRECON_PAGE_GUARD_UNBLOCK_SIGSEGV" with value "true"
[gfxrecon] INFO - Initializing GFXReconstruct capture layer
[gfxrecon] INFO -   GFXReconstruct Version 0.9.14-dev (panos_716:ecfc9ca)
[gfxrecon] INFO - Recording graphics API capture to gfxrecon_capture_20220711T130643.gfxr
[gfxrecon] INFO -   Vulkan Header Version 1.3.216
WARNING: lavapipe is not a conformant vulkan implementation, testing use only.

Test case 'KHR-Single-GL46.enhanced_layouts.glsl_constant_integral_expression'..
  Pass (Pass)

DONE!

Test run totals:
  Passed:        1/1 (100.0%)
  Failed:        0/1 (0.0%)
  Not supported: 0/1 (0.0%)
  Warnings:      0/1 (0.0%)
  Waived:        0/1 (0.0%)
```


I remeber being able to reproduce it back when you mention it but now I am working on a different pc and I cannot reproduce it.
Perhaps there's something wrong/different with my setup/libraries.

For reference here's the output from `LIBGL_ALWAYS_SOFTWARE=1 MESA_LOADER_DRIVER_OVERRIDE=zink glxgears -info`
```
$ LIBGL_ALWAYS_SOFTWARE=1 MESA_LOADER_DRIVER_OVERRIDE=zink glxgears -info
WARNING: lavapipe is not a conformant vulkan implementation, testing use only.
Running synchronized to the vertical refresh.  The framerate should be
approximately the same as the monitor refresh rate.
GL_RENDERER   = zink (llvmpipe (LLVM 13.0.1, 256 bits))
GL_VERSION    = 4.6 (Compatibility Profile) Mesa 22.1.3 (git-f0cda4094d)
GL_VENDOR     = Collabora Ltd
``` 
and `LIBGL_ALWAYS_SOFTWARE=1 MESA_LOADER_DRIVER_OVERRIDE=zink glxinfo`
[glxinfo.txt](https://github.com/LunarG/gfxreconstruct/files/9082941/glxinfo.txt)

**Comment by zmike** [comment]
I can confirm that the original issue is fixed, though I still get new crashes. I guess I'll file a new issue for that. Thanks!

**Comment by panos-lunarg** [comment]
Merge request in Mesa has been approved.
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/17273
